core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableDeltaTableScan.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.enumerable;
+
+import org.apache.calcite.linq4j.tree.BlockBuilder;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.schema.ModifiableTable;
+import org.apache.calcite.util.BuiltInMethod;
+
+
+/**
+ * Implementation of delta table (used to accumulate results in a recursive query) in
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}.
+ */
+public class EnumerableDeltaTableScan extends TableScan implements EnumerableRel {
```

Comment:
I wonder if this new subtype of `TableScan` would backfire when non-expecting rules would fire.
I think it would be better to have a standalone rel type.

Reply:
By looking into the code I don't see something fundamentally different from a regular table scan. Due to this I am less worried about non-expecting rules firing. However, I have the feeling that EnumerableDeltaTableScan and likewise LogicalDeltaTableScan can completely dissapear. LogicalDeltaTableScan can be possibly replaced by a regular LogicalTableScan and EnumerableDeltaTableScan by EnumerableTableScan. The logic of obtaining a table from the context could be part of the Table implementation.